### ISSUE TITLE
First-class support for luasockets (if installed) in mod API

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#10120]
## Problem

I maintain a few mods that require disabling mod security because they use lua sockets.  Using the new request_insecure_environment() _almost_ works, but because I'm using unix-domain sockets for the improved access control they offer, I can't even get the hacks that the IRC mod uses to work.

Mods that ONLY require sockets shouldn't need _full_ trust anyway.  Similar to the way HTTP works, it should be possible to require trust JUST for this feature an no other privileged access.

LuaSockets are one of the best options for IPC for things that are too complex or compute-bound to do in Lua inside MT, where some form of process barrier is desired, or where a resource is remote.
## Solutions

Add a way to configure mods as trusted for access to sockets ONLY, similar to how HTTP access works.  Load the LuaSockets module before constructing the mod security sandboxes so that it loads correctly (including all the tcp/udp/unix etc modules) and then provide a mod API for accessing them.  Ideally unix socket creation should be limited to paths that the mod would normally have access to via io.open et al.

## Alternatives

Emulating these operations may be possible in HTTP but it imposes a lot of overhead for cases where e.g. a persisted session with asynchronous message passing is desired.  Also this may require running additional middleware to convert HTTP requests back to raw socket operations for an external service.  HTTP also uses TCP for communication, which means that the only access control that's possible natively is host-based (i.e. binding to 127.0.0.1) and user-level access control requires more complexity and is vulnerable to more issues (e.g. MITM) that unix-domain sockets could prevent.

## Additional context

An example of a mod that needs something like this:
https://gitlab.com/sztest/szutilpack/-/tree/master/szutil_chatsocket

I tried to imitate what the IRC mod was doing with request_insecure_environment but it didn't work on Debian because LuaSocket is installed in a cpath that's specific to the system arch, release, version, etc.  Only by stealing the cpath back from an unsandboxed lua instance was I able to get it to work on Debian...
https://gitlab.com/sztest/szutilpack/-/blob/3b75077d2815ea511029ec818e200300adbd3169/szutil_chatsocket/init.lua#L45-65
...but then I wasn't able to get it to work on Alpine.

**Comment by appgurueu** [comment]
I fully agree. Another example (chat bridge as well) is https://github.com/appgurueu/adv_chat

**Comment by sfan5** [comment]
It sounds like the main factor for this feature request is the trouble you had when installing LuaSockets, but doing `ie.require("socket")` should work just fine *if* LuaSockets is installed exactly the way Minetest expects (easy to get wrong).
First-class support wouldn't really change this because the engine also has to rely on LuaSocket being installed correctly in your system.

**Comment by rubenwardy** [comment]
Could Minetest bundle luasockets? Does that even make sense?

**Comment by Warr1024** [comment]
> but doing ie.require("socket") should work just fine

...but ie.require("socket.unix") does not, and according to the luasockets documentation this was the only way I found to access unix-domain sockets.

> installed exactly the way Minetest expects

...which is also something that I don't have much control over, when it's controlled by a distro, and there are various reasons why I don't want to coopt what the distro provides and try to patch this myself.  Maintaining portability code is a PITA, MT already does a lot of work for this, and I wouldn't want to have to recreate or augment any of that if I'm going to have to maintain it myself and it won't be of use to any project other than mine.

> Could Minetest bundle luasockets?

I have no objection to this if it's the most expedient way to do it.  It seems like there could be some advantages to this, like not having to tell people "this mod requires a version of MT built with this optional component enabled."  Either way, if there were at least an obvious way to get it to work on each platform (e.g. download the binaries from this source on Windows, build with this cmake option on Linux) then that would be a big improvement.

**Comment by appgurueu** [comment]
A similar use case: Playing board games in Minetest (e.g. Chess or Go) against the computer. This can hardly be implemented sufficiently efficiently in Lua, and most engines are currently implemented in C(++).

**Comment by DanielHeath** [comment]
Another use case: I'd like to write a mod that runs an HTTP server which exposes a google-maps style view of the world - similar to https://unmined.net/example/1.18um/

**Comment by GreenXenith** [comment]
Properly including `socket.unix` would work if #10877 is resolved (the method provided by Desour in that thread works).
